### PR TITLE
PP-4614 Updates patch request to change individual `MerchantDetails`

### PR DESCRIPTION
**Pull request by cobainc0**
- adds tests for `ServiceResourceUpdate` and `ServiceUpdaterTest`
 - tests logic in `Service updater`
 - refactoring for consistency
 - removes unused `Nullable` annotation from `merchant_telephone_number`,
     `merchant_address_line2` and `merchant_email` in `MerchantDetailsEntity`,
     which seems to have been used to denote optional fields, however we do not
     enforce this on the DB

with @stephencdaly

### CODE REVIEW

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 1]
```diff
@@ -276,4 +272,94 @@ public void shouldUpdateCurrentGoLIveStageSuccessfully() {
...
+                .aServiceEntity()
+                .withMerchantDetailsEntity(null)
+                .build();
+
+        when(serviceDao.findByExternalId(serviceId)).thenReturn(Optional.of(serviceEntity));
+
+        Optional<Service> maybeService = updater.doUpdate(serviceId, serviceRequest);
+
+        assertThat(maybeService.isPresent(), is(true));
+        verify(serviceDao, times(1)).merge(serviceEntity);
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: Don’t need `times(1)`: it’s the default

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 2]
```diff
@@ -276,4 +272,94 @@ public void shouldUpdateCurrentGoLIveStageSuccessfully() {
...
+
+        when(serviceDao.findByExternalId(serviceId)).thenReturn(Optional.of(serviceEntity));
+
+        Optional<Service> maybeService = updater.doUpdate(serviceId, request);
+
+        assertThat(maybeService.isPresent(), is(true));
+        verify(serviceDao).merge(serviceEntity);
+        assertThat(maybeService.get().getMerchantDetails().getName(), is(name));
+        assertThat(maybeService.get().getMerchantDetails().getAddressLine1(), is(addressLine1));
+    }
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: Maybe this test could also assert the value of one of the unchanged fields just to confirm that they don’t get blown away

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java [hunk 3]
```diff
@@ -276,4 +272,94 @@ public void shouldUpdateCurrentGoLIveStageSuccessfully() {
...
+
+    private ServiceUpdateRequest getServiceUpdateRequest(String value, String path) {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: This isn’t really a getter, so I’d just name it `serviceUpdateRequest(…)` (and make it `static`!).

Tempting to take an `op` parameter too (even though it’s always `"replace"`) just for clarity.

Comment (cobainc0) [reply src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: 👍 updated

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java (hunk 2 above)
Comment (cobainc0) [reply src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: 👍 Nice one! I've amended this now, so we use`MerchantDetailsEntityBuilder` default values for merchant details fields. So, now on line: 341 we assert that the default name value is still "test-name" after updating addressLine1

File: src/test/java/uk/gov/pay/adminusers/resources/ServiceResourceUpdateMerchantDetailsTest.java [hunk 4]
```diff
@@ -109,4 +113,57 @@ public void shouldFail_whenUpdatingMerchantDetails_withMissingMandatoryFieldName
...
+
+        givenSetup()
+                .when()
+                .contentType(JSON)
+                .body(payload)
+                .patch(format(SERVICE_RESOURCE, serviceExternalId))
+                .then()
+                .statusCode(200)
+                .body("merchant_details.address_line1", is(addressLine1));
+    }
```
[the remark this answers is not in the document]
Comment (cobainc0) [reply src/test/java/uk/gov/pay/adminusers/resources/ServiceResourceUpdateMerchantDetailsTest.java]: As discussed we'll leave this here for now.

File: src/test/java/uk/gov/pay/adminusers/resources/ServiceResourceUpdateMerchantDetailsTest.java [hunk 5]
```diff
@@ -109,4 +113,57 @@ public void shouldFail_whenUpdatingMerchantDetails_withMissingMandatoryFieldName
                 .body("errors", hasItems("Field [name] is required"));
     }
 
+    @Test
+    public void shouldSucceed_whenPatchUpdatingMultipleMerchantDetails() {
+        String serviceExternalId = serviceDbFixture(databaseHelper).insertService().getExternalId();
+        String addressLine1 = "1 Spider Lane";
```
[the remark this answers is not in the document]
Comment (cobainc0) [reply src/test/java/uk/gov/pay/adminusers/resources/ServiceResourceUpdateMerchantDetailsTest.java]: 🕸

File: src/main/java/uk/gov/pay/adminusers/service/ServiceUpdater.java [hunk 6]
```diff
@@ -134,4 +143,53 @@ public Service doUpdateMerchantDetails(String serviceExternalId, UpdateMerchantD
...
+                getOrCreateMerchantDetails(serviceEntity).setEmail(serviceUpdateRequest.valueAsString());
+    }
+    
+    private BiConsumer<ServiceUpdateRequest, ServiceEntity> updateMerchantDetailsPhone() {
+        return (serviceUpdateRequest, serviceEntity) ->
+                getOrCreateMerchantDetails(serviceEntity).setTelephoneNumber(serviceUpdateRequest.valueAsString());
+    }
+    
+    private MerchantDetailsEntity getOrCreateMerchantDetails(ServiceEntity serviceEntity) {
+        if (serviceEntity.getMerchantDetailsEntity() != null)
```
[the remark this answers is not in the document]
Comment (cobainc0) [reply src/main/java/uk/gov/pay/adminusers/service/ServiceUpdater.java]: 👍 Thanks. Updated!

File: src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java (hunk 1 above)
Comment (cobainc0) [reply src/test/java/uk/gov/pay/adminusers/service/ServiceUpdaterTest.java]: 👍 removed. Thanks

**Pull request merged**